core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -273,6 +286,10 @@ private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation al
             // we only check for NO, since if this node is THROTTLING and it has enough "same data"
             // then we will try and assign it next time
             Decision decision = allocation.deciders().canAllocate(shard, node, allocation);
```

Comment:
The idea here was that, the only place these methods are currently called is from the allocator itself which has `explain=false`. When this is used from the explain API context, I was going to set debugDecision on the allocation in the beginning.  

However, you mentioning this made me realize that instead of passing `explain` to `makeAllocationDecision`, we can use allocation.debugDecision flag instead.  I'll make that change.